File: backend/app/services/file_parser_service.py
"""
文件解析服务

负责解析各种格式的文件并提取文本内容
支持：纯文本、Markdown、代码文件、PDF、Word 文档等
"""

import io
import logging
from pathlib import Path
from typing import Optional, Set
import aiofiles
import pdfplumber
from docx import Document

logger = logging.getLogger(__name__)


class FileParserService:
    """文件解析服务"""

    # 文件类型定义
    TEXT_EXTENSIONS: Set[str] = {
        "txt", "md", "markdown",  # 纯文本和 Markdown
        "py", "js", "ts", "jsx", "tsx",  # 代码文件
        "java", "cpp", "c", "h", "hpp",  # C/C++/Java
        "go", "rs", "rb", "php",  # 其他语言
        "json", "xml", "yaml", "yml", "toml",  # 配置文件
        "html", "htm", "css", "scss", "less",  # Web 文件
        "csv", "tsv",  # 数据文件
        "sql", "sh", "bat", "ps1",  # 脚本文件
    }
    
    PDF_EXTENSIONS: Set[str] = {"pdf"}
    WORD_EXTENSIONS: Set[str] = {"docx"}
    
    # 文件大小限制（字节）
    MAX_TEXT_SIZE: int = 10 * 1024 * 1024  # 10MB
    MAX_PDF_SIZE: int = 50 * 1024 * 1024  # 50MB
    MAX_WORD_SIZE: int = 20 * 1024 * 1024  # 20MB

    def __init__(self):
        """初始化文件解析服务"""
        pass

    async def detect_file_type(self, file_extension: str) -> str:
        """
        检测文件类型
        
        Args:
            file_extension: 文件扩展名（不含点，如 "pdf"）
        
        Returns:
            str: 文件类型（text, pdf, word, code, unknown）
        """
        ext = file_extension.lower().lstrip(".")
        
        if ext in self.TEXT_EXTENSIONS:
            return "text"
        elif ext in self.PDF_EXTENSIONS:
            return "pdf"
        elif ext in self.WORD_EXTENSIONS:
            return "word"
        elif ext in {"py", "js", "ts", "java", "cpp", "c", "go", "rs"}:
            return "code"
        else:
            return "unknown"

    async def parse_file(
        self,
        file_content: bytes,
        file_type: str,
        file_extension: str,
        file_size: Optional[int] = None,
    ) -> str:
        """
        解析文件内容
        
        Args:
            file_content: 文件内容（bytes）
            file_type: 文件类型（text/pdf/word/code）
            file_extension: 文件扩展名
            file_size: 文件大小（字节）
        
        Returns:
            str: 解析后的文本内容
        
        Raises:
            ValueError: 不支持的文件类型
            RuntimeError: 解析失败
        """
        ext = file_extension.lower().lstrip(".")
        
        # 验证文件大小
        if file_size:
            await self._validate_file_size(file_size, file_type)
        
        try:
            if file_type == "text" or file_type == "code":
                return await self._parse_text_file(file_content, ext)
            elif file_type == "pdf":
                return await self._parse_pdf_file(file_content)
            elif file_type == "word":
                return await self._parse_word_file(file_content)
            else:
                # 未知类型，尝试当作文本解析
                logger.warning(f"未知文件类型：{file_type}，尝试当作文本解析")
                return await self._parse_text_file(file_content, ext)
                
        except Exception as e:
            logger.error(f"文件解析失败：{file_extension}, size={file_size}, error={e}")
            raise RuntimeError(f"文件解析失败：{str(e)}")

    async def _validate_file_size(self, file_size: int, file_type: str) -> None:
        """验证文件大小"""
        limits = {
            "text": self.MAX_TEXT_SIZE,
            "code": self.MAX_TEXT_SIZE,
            "pdf": self.MAX_PDF_SIZE,
            "word": self.MAX_WORD_SIZE,
        }
        
        limit = limits.get(file_type, self.MAX_TEXT_SIZE)
        if file_size > limit:
            raise ValueError(f"文件超出大小限制（最大 {limit // 1024 // 1024}MB）")

    async def _parse_text_file(self, content: bytes, ext: str) -> str:
        """
        解析文本文件
        
        Args:
            content: 文件内容
            ext: 文件扩展名
        
        Returns:
            str: 文本内容
        """
        # 尝试不同的编码
        encodings = ["utf-8", "gbk", "gb2312", "latin-1"]
        
        for encoding in encodings:
            try:
                text = content.decode(encoding)
                # 清理空白字符，但保留原有结构
                return text.strip()
            except UnicodeDecodeError:
                continue
        
        # 如果所有编码都失败，使用 latin-1（不会抛出异常）
        return content.decode("latin-1", errors="replace").strip()

    async def _parse_pdf_file(self, content: bytes) -> str:
        """
        解析 PDF 文件
        
        Args:
            content: PDF 文件内容
        
        Returns:
            str: 提取的文本内容
        """
        text_parts = []
        
        try:
            with pdfplumber.open(io.BytesIO(content)) as pdf:
                for page in pdf.pages:
                    page_text = page.extract_text()
                    if page_text:
                        text_parts.append(page_text)
            
            return "\n\n".join(text_parts).strip()
            
        except Exception as e:
            logger.error(f"PDF 解析失败：{e}")
            raise RuntimeError(f"PDF 文件解析失败：{str(e)}")

    async def _parse_word_file(self, content: bytes) -> str:
        """
        解析 Word 文档（.docx）
        
        Args:
            content: Word 文件内容
        
        Returns:
            str: 提取的文本内容
        """
        try:
            doc = Document(io.BytesIO(content))
            paragraphs = [para.text.strip() for para in doc.paragraphs if para.text.strip()]
            return "\n\n".join(paragraphs)
            
        except Exception as e:
            logger.error(f"Word 文档解析失败：{e}")
            raise RuntimeError(f"Word 文档解析失败：{str(e)}")

    async def parse_file_from_path(self, file_path: str) -> str:
        """
        从文件路径解析内容
        
        Args:
            file_path: 文件路径
        
        Returns:
            str: 解析后的文本内容
        """
        path = Path(file_path)
        ext = path.suffix.lstrip(".").lower()
        file_type = await self.detect_file_type(ext)
        
        async with aiofiles.open(file_path, "rb") as f:
            content = await f.read()
        
        return await self.parse_file(
            file_content=content,
            file_type=file_type,
            file_extension=ext,
            file_size=len(content),
        )

    def get_supported_extensions(self) -> dict:
        """
        获取支持的文件扩展名列表
        
        Returns:
            dict: {文件类型：[扩展名列表]}
        """
        return {
            "text": sorted(list(self.TEXT_EXTENSIONS)),
            "pdf": sorted(list(self.PDF_EXTENSIONS)),
            "word": sorted(list(self.WORD_EXTENSIONS)),
            "code": ["py", "js", "ts", "java", "cpp", "c", "go", "rs"],
        }

File: backend-ts/src/modules/knowledge-base/file-parser.service.ts
import { Injectable, Logger } from "@nestjs/common";
import * as fs from "fs";
import * as path from "path";
import { PDFParse } from "pdf-parse";
import * as mammoth from "mammoth";

@Injectable()
export class FileParserService {
  private readonly logger = new Logger(FileParserService.name);

  // 支持的文件类型
  private readonly TEXT_EXTENSIONS = new Set([
    "txt",
    "md",
    "markdown",
    "py",
    "js",
    "ts",
    "jsx",
    "tsx",
    "java",
    "cpp",
    "c",
    "h",
    "hpp",
    "go",
    "rs",
    "rb",
    "php",
    "json",
    "xml",
    "yaml",
    "yml",
    "toml",
    "html",
    "htm",
    "css",
    "scss",
    "less",
    "csv",
    "tsv",
    "sql",
    "sh",
    "bat",
    "ps1",
  ]);
  private readonly PDF_EXTENSIONS = new Set(["pdf"]);
  private readonly WORD_EXTENSIONS = new Set(["docx"]);

  // 文件大小限制（字节）
  private readonly MAX_TEXT_SIZE = 10 * 1024 * 1024; // 10MB
  private readonly MAX_PDF_SIZE = 50 * 1024 * 1024; // 50MB
  private readonly MAX_WORD_SIZE = 20 * 1024 * 1024; // 20MB

  /**
   * 检测文件类型
   */
  async detectFileType(fileExtension: string): Promise<string> {
    const ext = fileExtension.toLowerCase().replace(/^\./, "");

    if (this.TEXT_EXTENSIONS.has(ext)) {
      return "text";
    } else if (this.PDF_EXTENSIONS.has(ext)) {
      return "pdf";
    } else if (this.WORD_EXTENSIONS.has(ext)) {
      return "word";
    } else if (
      ["py", "js", "ts", "java", "cpp", "c", "go", "rs"].includes(ext)
    ) {
      return "code";
    } else {
      return "unknown";
    }
  }

  /**
   * 解析文件内容
   */
  async parseFile(
    fileContent: Buffer,
    fileType: string,
    fileExtension: string,
    fileSize?: number,
  ): Promise<string> {
    const ext = fileExtension.toLowerCase().replace(/^\./, "");

    // 验证文件大小
    if (fileSize) {
      this.validateFileSize(fileSize, fileType);
    }

    try {
      if (fileType === "text" || fileType === "code") {
        return await this.parseTextFile(fileContent, ext);
      } else if (fileType === "pdf") {
        return await this.parsePdfFile(fileContent);
      } else if (fileType === "word") {
        return await this.parseWordFile(fileContent);
      } else {
        // 未知类型，尝试当作文本解析
        this.logger.warn(`未知文件类型：${fileType}，尝试当作文本解析`);
        return await this.parseTextFile(fileContent, ext);
      }
    } catch (error: any) {
      this.logger.error(
        `文件解析失败：${fileExtension}, size=${fileSize}, error=${error.message}`,
      );
      throw new Error(`文件解析失败：${error.message}`);
    }
  }

  /**
   * 验证文件大小
   */
  private validateFileSize(fileSize: number, fileType: string): void {
    const limits: Record<string, number> = {
      text: this.MAX_TEXT_SIZE,
      code: this.MAX_TEXT_SIZE,
      pdf: this.MAX_PDF_SIZE,
      word: this.MAX_WORD_SIZE,
    };

    const limit = limits[fileType] || this.MAX_TEXT_SIZE;
    if (fileSize > limit) {
      throw new Error(
        `文件超出大小限制（最大 ${Math.floor(limit / 1024 / 1024)}MB）`,
      );
    }
  }

  /**
   * 解析文本文件
   */
  private async parseTextFile(content: Buffer, ext: string): Promise<string> {
    // 尝试不同的编码
    const encodings = ["utf-8", "gbk", "gb2312", "latin-1"];

    for (const encoding of encodings) {
      try {
        const text = content.toString(encoding as BufferEncoding);
        return text.trim();
      } catch (error) {
        continue;
      }
    }

    // 如果所有编码都失败，使用 latin-1（不会抛出异常）
    return content.toString("latin1" as BufferEncoding).trim();
  }

  /**
   * 解析 PDF 文件
   */
  private async parsePdfFile(content: Buffer): Promise<string> {
    try {
      const parser = new PDFParse({ data: content });
      const result = await parser.getText();
      // 清理文本：去除多余空白、标准化换行符
      return result.text
        .replace(/\r\n/g, "\n")
        .replace(/\r/g, "\n")
        .replace(/\n{3,}/g, "\n\n")
        .trim();
    } catch (error: any) {
      this.logger.error(`PDF 解析失败: ${error.message}`);
      throw new Error(`PDF 文件解析失败: ${error.message}`);
    }
  }

  /**
   * 解析 Word 文档
   */
  private async parseWordFile(content: Buffer): Promise<string> {
    try {
      this.logger.debug("开始解析 Word 文档");
      // 优先提取纯文本
      const result = await mammoth.extractRawText({ buffer: content });

      if (result.value && result.value.trim()) {
        this.logger.debug("Word 文档解析成功");
        // 清理文本：标准化换行符、合并多余空行、去除首尾空白
        return result.value
          .replace(/\r\n/g, "\n")
          .replace(/\r/g, "\n")
          .replace(/\n{3,}/g, "\n\n")
          .trim();
      }

      // 如果纯文本为空，记录警告并返回空字符串或尝试其他方案
      this.logger.warn("Word 文档提取的纯文本内容为空");
      return "";
    } catch (error: any) {
      this.logger.error(`Word 文档解析失败: ${error.message}`);
      throw new Error(`Word 文件解析失败: ${error.message}`);
    }
  }

  /**
   * 从文件路径解析内容
   */
  async parseFileFromPath(filePath: string): Promise<string> {
    const ext = path.extname(filePath).replace(/^\./, "").toLowerCase();
    const fileType = await this.detectFileType(ext);

    const content = await fs.promises.readFile(filePath);

    return await this.parseFile(content, fileType, ext, content.length);
  }

  /**
   * 获取支持的文件扩展名列表
   */
  getSupportedExtensions(): Record<string, string[]> {
    return {
      text: Array.from(this.TEXT_EXTENSIONS).sort(),
      pdf: Array.from(this.PDF_EXTENSIONS).sort(),
      word: Array.from(this.WORD_EXTENSIONS).sort(),
      code: ["py", "js", "ts", "java", "cpp", "c", "go", "rs"],
    };
  }
}
